Mathlib/RingTheory/EssentialFiniteness.lean

```diff
@@ -0,0 +1,211 @@
+  ((IsUnit.submonoid S).comap (algebraMap (EssFiniteType.subalgebra R S) S))
+
+instance EssFiniteType.isLocalization [h : EssFiniteType R S] :
+    IsLocalization (EssFiniteType.submonoid R S) S :=
+  h.cond.choose_spec
+
+variable {R S} in
+lemma EssFiniteType.algHom_ext [EssFiniteType R S]
+    (f g : S →ₐ[R] T) (H : ∀ s ∈ finset R S, f s = g s) : f = g := by
+  suffices f.toRingHom = g.toRingHom by ext; exact RingHom.congr_fun this _
+  apply IsLocalization.ringHom_ext (EssFiniteType.submonoid R S)
+  suffices f.comp (IsScalarTower.toAlgHom R _ S) = g.comp (IsScalarTower.toAlgHom R _ S) by
+    ext; exact AlgHom.congr_fun this _
+  apply AlgHom.ext_of_adjoin_eq_top (s := { x | x.1 ∈ finset R S })
+  · rw [← top_le_iff]
+    rintro x _
+    refine' Algebra.adjoin_induction' _ _ _ _ x
+    · intro x hx; exact Algebra.subset_adjoin hx
+    · intro r; exact Subalgebra.algebraMap_mem _ _
+    · intro x y hx hy; exact add_mem hx hy
+    · intro x y hx hy; exact mul_mem hx hy
+  · rintro ⟨x, hx⟩ hx'; exact H x hx'
+
+instance EssFiniteType.baseChange [h : EssFiniteType R S] : EssFiniteType T (T ⊗[R] S) := by
+  classical
+  rw [essFiniteType_iff] at h ⊢
+  obtain ⟨σ, hσ⟩ := h
+  use σ.image Algebra.TensorProduct.includeRight
+  intro s
+  induction' s using TensorProduct.induction_on with x y x y hx hy
+  · exact ⟨1, one_mem _, isUnit_one, by simpa using zero_mem _⟩
+  · obtain ⟨t, h₁, h₂, h₃⟩ := hσ y
+    have H : ∀ x : S, x ∈ Algebra.adjoin R (σ : Set S) →
+        1 ⊗ₜ[R] x ∈ Algebra.adjoin T
+          ((σ.image Algebra.TensorProduct.includeRight : Finset (T ⊗[R] S)) : Set (T ⊗[R] S)) := by
+      intro x hx
+      have : Algebra.TensorProduct.includeRight x ∈
+          (Algebra.adjoin R (σ : Set S)).map (Algebra.TensorProduct.includeRight (A := T)) :=
+        Subalgebra.mem_map.mpr ⟨_, hx, rfl⟩
+      rw [← Algebra.adjoin_adjoin_of_tower R]
+      apply Algebra.subset_adjoin
+      simpa [← Algebra.adjoin_image] using this
+    refine ⟨Algebra.TensorProduct.includeRight t, H _ h₁, h₂.map _, ?_⟩
+    simp only [Algebra.TensorProduct.includeRight_apply, Algebra.TensorProduct.tmul_mul_tmul,
+      mul_one]
+    rw [← mul_one x, ← smul_eq_mul, ← TensorProduct.smul_tmul']
+    apply Subalgebra.smul_mem
+    exact H _ h₃
+  · obtain ⟨tx, hx₁, hx₂, hx₃⟩ := hx
+    obtain ⟨ty, hy₁, hy₂, hy₃⟩ := hy
+    refine ⟨_, mul_mem hx₁ hy₁, hx₂.mul hy₂, ?_⟩
+    rw [add_mul, ← mul_assoc, mul_comm tx ty, ← mul_assoc]
+    exact add_mem (mul_mem hx₃ hy₁) (mul_mem hy₃ hx₁)
+
+lemma EssFiniteType.of_comp [h : EssFiniteType R T] : EssFiniteType S T := by
+  rw [essFiniteType_iff] at h ⊢
+  classical
+  obtain ⟨σ, hσ⟩ := h
+  use σ
+  intro x
+  obtain ⟨y, hy₁, hy₂, hy₃⟩ := hσ x
+  simp_rw [← Algebra.adjoin_adjoin_of_tower R (S := S) (σ : Set T)]
+  exact ⟨y, Algebra.subset_adjoin hy₁, hy₂, Algebra.subset_adjoin hy₃⟩
+
+lemma EssFiniteType.of_comp_iff [EssFiniteType R S] :
+    EssFiniteType R T ↔ EssFiniteType S T :=
```

Comment:
Should this be called `EssFiniteType.comp_iff`? (`of_comp` indicates an hypothesis to me).